api/src/main/java/org/apache/iceberg/util/WrapperSet.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.util;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterators;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+
+/**
+ * A custom set for a {@link Wrapper} of the given type that maintains insertion order
+ *
+ * @param <T> The type to wrap in a {@link Wrapper} instance.
+ */
+abstract class WrapperSet<T> implements Set<T>, Serializable {
+  private final Set<Wrapper<T>> set;
+
+  protected WrapperSet(Iterable<Wrapper<T>> wrappers) {
+    this.set = Sets.newLinkedHashSet(wrappers);
+  }
+
+  protected abstract Wrapper<T> wrapper();
+
+  protected abstract Wrapper<T> wrap(T file);
+
+  protected abstract boolean isInstance(Object obj);
+
+  protected interface Wrapper<T> {
+    T get();
+
+    Wrapper<T> set(T object);
+  }
+
+  @Override
+  public int size() {
+    return set.size();
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return set.isEmpty();
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public boolean contains(Object obj) {
+    if (isInstance(obj)) {
+      Wrapper<T> wrapper = wrapper();
+      boolean result = set.contains(wrapper.set((T) obj));
+      wrapper.set(null); // don't hold a reference to the value
+      return result;
+    }
+
+    return false;
+  }
+
+  @Override
+  public Iterator<T> iterator() {
+    return Iterators.transform(set.iterator(), Wrapper::get);
+  }
+
+  @Override
+  public Object[] toArray() {
+    return Lists.newArrayList(iterator()).toArray();
+  }
+
+  @Override
+  public <X> X[] toArray(X[] destArray) {
+    return Lists.newArrayList(iterator()).toArray(destArray);
+  }
+
+  @Override
+  public boolean add(T obj) {
+    return set.add(wrap(obj));
```

Comment:
I suspect that we can add `null` but not retrieve it back. It is up to us whether we want to support nulls but it has to be consistent. It is valid to throw `NullPointerException` if the set does not support nulls.

Reply:
you're right and I don't think we want to support nulls inside the set. I've updated the impl to throw a NPE when adding/removing nulls or when a null collection is passed. I was initially aligning the impl fully with `CharSequenceSet` but `CharSequenceSet` has the same inconsistency issue